docs/docs/higher-order-components.md

```diff
@@ -0,0 +1,338 @@
+# Higher-order components
+
+A higher-order component (HOC) is an advanced technique in React for reusing component logic.
+
+Concretely, **a higher-order component is a function that returns a component.** An HOC accepts a component as one of its arguments, and returns an enhanced version of that component:
+
+```js
+let EnhancedComponent = higherOrderComponent(ChildComponent)
+```
+
+Whereas a component transforms props into UI, a higher-order component transforms one component into another component.
+
+HOCs are common in third-party React libraries, such as Redux's `connect()` and Relay's `createContainer()`.
+
+In this document, we'll discuss why higher-order components are useful, and how to write your own.
+
+## Use HOCs for cross-cutting concerns
+
+Components are the primary unit of code reuse in React. However, you'll find that some patterns aren't easily abstracted into traditional components.
+
+For example, say you have a CommentList component that subscribes to an external data source to render a list of comments:
+
+```js
+class CommentList extends React.Component {
+  constructor(props) {
+    super();
+    this.handleChange = this.handleChange.bind(this);
+    this.state = {
+      // Store data in state
+      comments: DataSource.getComments()
+    };
+  }
+
+  componentDidMount() {
+    // Subscribe to changes
+    DataSource.addChangeListener(this.handleChange);
+  }
+
+  componentWillUnmount() {
+    // Clean up listener
+    DataSource.removeChangeListener(this.handleChange);
+  }
+
+  handleChange() {
+    // Update component state whenever the data source changes
+    this.setState({
+      comments: DataSource.getComments()
+    });
+  }
+
+  render() {
+    return (
+      <div>
+        {comments.map(function(comment) {
```

Comment:
Let's just use arrows here, I plan to use them in such cases throughout the docs